### ISSUE TITLE
Highlighting breaks if no semicolon after type expressions

### ISSUE THREAD

**Issue by StreetStrider** [babel/babel-sublime#309]
When using something from Flow syntax, highlighting may break if no semicolons used.
For sample, put something like:
```typescript
export type A = string
type B = number
```
The second one would not be properly highlighted if the first one losses semicolon at the end. However, the syntax is correct.

My linting rules encourages to evade semicolons where possible, so I can't use semicolon at the end of the line.

`import type` statements works good.
This would be highlighted properly:
```typescript
import type { A } from './A'
```

**Comment by bdefore** [comment]
I've also encountered this. Interesting that the `type` declaration would have any impact on semicolons.

**Comment by bdefore** [comment]
See also #298

**Comment by mareksuscak** [comment]
Not sure if it's related but same for static class properties:

<img width="234" alt="screen shot 2017-04-14 at 01 12 54" src="https://cloud.githubusercontent.com/assets/613647/25027778/8a00c45c-20af-11e7-9a7d-efb5a3c7bdb9.png">

```js
class Component extends React.Component {
  static propTypes = {
    config: PropTypes.object
  }

  static defaultProps = {
    config: {}
  }
}
```

**Comment by StreetStrider** [comment]
@mareksuscak seemes related

**Comment by JemarJones** [comment]
Same issues as @mareksuscak

**Comment by miyabi** [comment]
Same as @mareksuscak.
It is fixed by inserting semicolon after closing brace, but it is unwanted.

**Comment by Xazzzi** [comment]
You are welcome to test if it works for you.

**Comment by michaelcarter** [comment]
@Xazzzi I've googled around but am struggling, how can I install this fixed version in Sublime 3?

**Comment by mareksuscak** [comment]
@michaelcarter It hasn't been released yet. So you have to go the `~/Library/Application Support/Sublime Text 3/Packages/Babel` and edit the file `JavaScript (Babel).sublime-syntax` accordingly:

<img width="428" alt="screen shot 2017-05-15 at 14 02 12" src="https://cloud.githubusercontent.com/assets/613647/26056808/28211670-3977-11e7-9c1a-54c907bed85b.png">

**Comment by michaelcarter** [comment]
@mareksuscak This is all I have in `~/Library/Application Support/Sublime Text 3/Packages/`:

<img width="478" alt="screen shot 2017-05-16 at 14 41 30" src="https://cloud.githubusercontent.com/assets/1911741/26108969/ea1604b0-3a45-11e7-80f2-22dda9d45ba3.png">

**Comment by Xazzzi** [comment]
Unfortunately i don't own any apple pc, so i can't really tell how to fix it on Mac.
What i suggest though is to try modifying `JavaScript (Babel).tmLanguage` in a similar way.

**Comment by JemarJones** [comment]
@Xazzzi It does not work for me. Instead the issue happens for all declarations in the class (after the contractor) instead of just the one.

**Comment by mareksuscak** [comment]
@michaelcarter that's odd, here's mine:

<img width="447" alt="screen shot 2017-05-17 at 00 05 19" src="https://cloud.githubusercontent.com/assets/613647/26130348/95e70264-3a94-11e7-9fd4-dcbfeba46376.png">

How did you install babel package? Via [Package Control](https://packagecontrol.io/installation), right? I use `Babel` package `v8.6.3`. Make sure you use the latest version.

<img width="418" alt="screen shot 2017-05-17 at 00 04 54" src="https://cloud.githubusercontent.com/assets/613647/26130397/cb4418c0-3a94-11e7-916a-af9b90fe3a7c.png">

**Comment by michaelcarter** [comment]
@mareksuscak @Xazzzi Ok, I've got the file there now but editing it does nothing to improve my syntax highlighting. I'm left with code like this:

<img width="544" alt="screen shot 2017-05-17 at 10 30 53" src="https://cloud.githubusercontent.com/assets/1911741/26147597/f7034c72-3aeb-11e7-90db-c3b764c73235.png">

Drop in a `;` at the end of the type declaration and I get this (much better!):

<img width="551" alt="screen shot 2017-05-17 at 10 31 41" src="https://cloud.githubusercontent.com/assets/1911741/26147635/101fb7ae-3aec-11e7-8681-d1bc0e8d4b38.png">

Thanks for all your help so far.

**Comment by mareksuscak** [comment]
@michaelcarter Have you got it working yet? I did not have much time to dig in. I'm currently not very familiar with this project and there doesn't seem to be a whole lot of documentation (i.e. what's the preferred way to convert `*.tmLanguage` to `*.sublime-syntax`). 

One thing I can say for sure is that it doesn't work for you because @Xazzzi's solution only fixes missing semicolon for class properties. Your code is breaking on `flowtype-declaration` so a similar patch is necessary for those.

**Comment by michaelcarter** [comment]
@mareksuscak No. I'm afraid not. I've tried modifying the flowtype-declaration section of my sublime-syntax document but have had no success in getting something working (I have no idea what I'm doing, so that's unsurprising).

I don't usually work with flow, so I'll probably just hop over to Atom for this project so I have something that works and then check back periodically on whether this is working in newer versions.

Thanks for all your help! 🙏

**Comment by sankarravi** [comment]
I got this working using @Xazzzi's rough change in the `flowtype-declaration` section. Notably, I had to the change the `sublime-syntax` file, not the `tmLanguage` file:
`~/Library/Application Support/Sublime Text 3/Packages/Babel/JavaScript (Babel).sublime-syntax`

My final flowtype-declaration looks like:
```
flowtype-declaration:
  - match: (?<!\.)\b(declare)\b
    scope: support.type.declare.flowtype.js

  - match: (?<!\.)\b(type)\b(?=\s*[_$a-zA-Z])
    scope: support.type.type.flowtype.js
    push:
      - meta_scope: meta.type.flowtype.js
      - match: (?=\s*(;|from))
        pop: true
      - include: flowtype-tokens
      - match: $
        pop: true

  - match: (?<!\.)\b(type)\b(?=\s*{)
    scope: support.type.type.flowtype.js
    push:
      - meta_scope: meta.type.flowtype.js
      - match: (?<=})
        pop: true
      - include: flowtype-brackets
      - include: flowtype-polymorph
      - match: ([_$a-zA-Z][$\w]*)
        scope: entity.name.type.js
      - match: $
        pop: true
```
starting at L207

Screenshot:
![image](https://cloud.githubusercontent.com/assets/1098507/26807736/146c63fe-4a0d-11e7-921b-ccefff8a9f48.png)

**Comment by Mottoweb** [comment]
Thanks @Xazzzi @sankarravi, this has to be a PR.
I almost switched to atom, but then found this.

**Comment by amsul** [comment]
@Xazzzi @sankarravi you two are legends! Saved me the effort of switching to vscode 😅

**Comment by StreetStrider** [comment]
Can someone help on converting to `tmLanguage`? I'd found analogous definition in tmLanguage file and I'm trying to port changes.
Looks like the key is:
```
- match: $
    pop: true
```
However, I can't find analogue in tmLanguage. Maybe it's encoded in some combination of `beginCaptures`/`begin` indirectly?
I haven't got any experience in theese DSLs, so I would glad to get any help.

**Comment by breadadams** [comment]
_NICE_ @sankarravi @Xazzzi 👏 

Any word on getting this merged into a release?

**Comment by Xazzzi** [comment]
I wonder if it could be merged. I dont have any experience in language defs DSL's either and found this fix by trying different things (and failing a lot).
This repo seems pretty much abandoned, as last update was 9 month ago.
We can ask @zertosh maybe ?

**Comment by StreetStrider** [comment]
It would be awesome if someone ports this change to `tmLanguage` as well. I use babel-sublime `tmLanguage` with my own fixes in conjunction with my own theme. I tried to do port myself, but have not enough experience with theese themes, failed to do it in simple manner. It's much to ask, but it would be cool to see this happen.

**Comment by borela** [comment]
For sublime, try FJSX15 https://github.com/borela/naomi

**Comment by emwalker** [comment]
Curious what the status of this is.  In an [earlier PR](https://github.com/babel/babel-sublime/pull/322) by @Xazzzi, the fix was place din the `tmLanguage` file, and that seems to have caused problems. In @sankarravi's [comment above](https://github.com/babel/babel-sublime/issues/309#issuecomment-306340132), by contrast, the fix was placed in the `sublime-syntax` file, and this worked for me and at least one other person without causing any obvious issues.

This bug makes it hard to want to use flow if your repo doesn't use semicolons, and the fix seems to be an easy one.   @zertosh, are there further considerations we should be aware of, or would you be open to a PR?

**Comment by webbuyrich** [comment]
After installing Babel in sublime I followed the readme instructions in the package control webpage
[https://packagecontrol.io/packages/Babel](https://packagecontrol.io/packages/Babel)

Setting as the default syntax
To set it as the default syntax for a particular extension:

1. Open a file with that extension
2. Select View from the menu
3. Then Syntax -> Open all with current extension as... -> Babel -> JavaScript (Babel). 
4. Repeat this for each extension (e.g.: .js and .jsx)

**Comment by emwalker** [comment]
Does anyone know of a fork of this repo that is up to date and being maintained?

**Comment by StreetStrider** [comment]
@emwalker use Naomi (above in thread).

**Comment by Thom1729** [comment]
This should be fixed in v10.0.0-beta.1. See #379 for more information on the new release. Please report any problems with v10 in a new issue.
